runtime/doc/lua.txt

```diff
@@ -1581,4 +1581,70 @@ uri_to_fname({uri})                                       *vim.uri_to_fname()*
                 Return: ~
                     Filename
 
+==============================================================================
+Lua module: diff                                                    *lua-diff*
+
+diff({a}, {b}, {opts})                                            *vim.diff()*
+                Run diff on strings {a} and {b}.
+
+                Examples: >
+                vim.diff('a\n', 'b\nc\n')
+                -->
+                @@ -1 +1,2 @@
+                -a
+                +b
+                +c
+
+                vim.diff('a\n', 'b\nc\n', {hunk_lines = true})
+                -->
+                {
+                    {1, 1, 1, 2}
+                }
+<
+                Parameters: ~
+                    {a}      First string to compare
+                    {b}      Second string to compare
+                    {opts}   Optional parameters:
+                             • `on_hunk` (callback):
+                               Invoked for each hunk in the diff. Return a
+                               negative number to cancel the callback for any
+                               remaining hunks.
+                               Args:
+                               • `start_a` (integer): Start line of hunk in {a}, 1-based.
+                               • `count_a` (integer): Hunk size in {a}.
+                               • `start_b` (integer): Start line of hunk in {b}, 1-based.
+                               • `count_b` (integer): Hunk size in {b}.
+                             • `hunk_lines` (boolean):
+                               Return an array of the hunk locations. Cannot
+                               be used with `on_hunk`.
```

Comment:
Future idea: would it make sense to optionally return a structured result, as a map instead of a unified diff string?

Then to leave the door open for that, we probably want a single parameter (string instead of bool) than controls the result form. Something like

    ```suggestion
                             • `result_type` (string): Controls the form of the returned diff. Cannot be used with `on_hunk`.
                               • "string": (default) unified diff string 
                               • "lines": array of the hunk locations
                               • "map": (not implemented) structured diff in the form { ... }
```

Reply:
maybe "unified" instead of "strings", and not mention "map", as it is no reason to assume that a future structured format must be a map?

Reply:
Hmm, for the unified option, do we want to split the strings to be consistent with nvim_buf_get_lines?

Also what about `indices` instead of `lines`?

Reply:
let's not split the strings. it is not certain that you would put them directly in a buffer, or something.

Reply:
Sure, I'll take the less work. Just thought split strings are much easier to work with and can be more easily iterated over.

Reply:
In generally lua has no problem dealing with multiline strings (unlike vimL, where a string cannot represent buffer contents). if the user has a usecase, they can split the string themselves (and not necessarily on just `\n`, just as well per hunk or something).